pkg/query-service/app/clickhouseReader/reader.go

```diff
@@ -2753,3 +2765,266 @@ func (r *ClickHouseReader) GetSamplesInfoInLastHeartBeatInterval(ctx context.Con
+		err := r.db.Exec(ctx, query)
+		if err != nil {
+			return &model.ApiError{Err: err, Typ: model.ErrorInternal}
+		}
+	}
+	return nil
+}
+
+func (r *ClickHouseReader) GetLogs(ctx context.Context, params *model.LogsFilterParams) (*[]model.GetLogsResponse, *model.ApiError) {
+	response := []model.GetLogsResponse{}
+	fields, apiErr := r.GetLogFields(ctx)
+	if apiErr != nil {
+		return nil, apiErr
+	}
+
+	filterSql, err := logs.GenerateSQLWhere(fields, params)
+	if err != nil {
+		return nil, &model.ApiError{Err: err, Typ: model.ErrorBadData}
+	}
+
+	query := fmt.Sprintf("%s from %s.%s", constants.LogsSQLSelect, r.logsDB, r.logsTable)
+
+	if filterSql != "" {
+		query += fmt.Sprintf(" where %s", filterSql)
+	}
+
+	query = fmt.Sprintf("%s order by %s %s limit %d", query, params.OrderBy, params.Order, params.Limit)
+	zap.S().Debug(query)
+	err = r.db.Select(ctx, &response, query)
+	if err != nil {
+		return nil, &model.ApiError{Err: err, Typ: model.ErrorInternal}
+	}
+
+	return &response, nil
+}
+
+func (r *ClickHouseReader) TailLogs(ctx context.Context, client *model.LogsTailClient) {
+
+	fields, apiErr := r.GetLogFields(ctx)
+	if apiErr != nil {
+		client.Error <- apiErr.Err
+		return
+	}
+
+	filterSql, err := logs.GenerateSQLWhere(fields, &model.LogsFilterParams{
+		Query: client.Filter.Query,
+	})
+
+	if err != nil {
+		client.Error <- err
+		return
+	}
+
+	query := fmt.Sprintf("%s from %s.%s", constants.LogsSQLSelect, r.logsDB, r.logsTable)
+
+	tsStart := uint64(time.Now().UnixNano())
+	if client.Filter.TimestampStart != 0 {
+		tsStart = client.Filter.TimestampStart
+	}
+
+	var idStart string
+	if client.Filter.IdStart != "" {
+		idStart = client.Filter.IdStart
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			done := true
+			client.Done <- &done
+			zap.S().Debug("closing go routine : " + client.Name)
+			return
+		default:
+			// get the new 100 logs as anything more older won't make sense
+			tmpQuery := fmt.Sprintf("%s where timestamp >='%d'", query, tsStart)
+			if filterSql != "" {
+				tmpQuery += fmt.Sprintf(" and %s", filterSql)
+			}
+			if idStart != "" {
+				tmpQuery += fmt.Sprintf(" and id > '%s'", idStart)
+			}
+			tmpQuery = fmt.Sprintf("%s order by timestamp desc, id desc limit 100", tmpQuery)
+			zap.S().Debug(tmpQuery)
+			response := []model.GetLogsResponse{}
+			err := r.db.Select(ctx, &response, tmpQuery)
+			if err != nil {
+				zap.S().Debug(err)
+				client.Error <- err
+				return
+			}
+			len := len(response)
+			for i := len - 1; i >= 0; i-- {
+				select {
+				case <-ctx.Done():
+					done := true
+					client.Done <- &done
+					zap.S().Debug("closing go routine while sending logs : " + client.Name)
+					return
+				default:
+					client.Logs <- &response[i]
+					if i == 0 {
+						tsStart = response[i].Timestamp
+						idStart = response[i].ID
+					}
+				}
+			}
+			time.Sleep(time.Duration(r.liveTailRefreshSeconds) * time.Second)
```

Comment:
Ticker seems like a good fit here. Was there any reason not to use `Ticker` and instead choose `Sleep`?

Reply:
Didn't know about this, let me go through it.

Reply:
thanks, have updated it to a ticker.